### ISSUE TITLE
[Bug] Game sometimes crashes when FX is being held

### ISSUE THREAD

**Issue by 123jimin** [Drewol/unnamed-sdvx-clone#394]
The crash happenes very rarely during manual play, but it happens often during autoplay. TapeStop FX effects are most prone to this crash.

I have found the reason, and I believe that this is also the reason behind occasional audio glitch. However, the fix is quite *annoying* to do and I wanted to rant on how the structure can be improved....

The cause is a common kind of race condition.
1. DSP processing is handled in the audio thread; (on Windows) the main function is `AudioOutput_Impl::AudioThread`.
2. When a new effect is being added (ex: by pressing FX), `GameAudioEffect::CreateDSP` is called from `AudioPlayback`, which is run on the main thread (`Application::m_MainLoop`).
3. When DSPs are added to the audio track, it is done via `AudioBase::AddDSP`, which is thread-safe.
4. However, DSP parameters are set *after* they are added to the audio track, and it's done on the main thread!
5. This can cause *a lot* of problem, but the most significant one is the crash; for example `TapeStopDSP::SetLength` clears `m_sampleBuffer` and this can happen anytime during `TapeStopDSP::Process`. It will cause out-of-bound crash then.

The simplest solution is setting the parameters before calling `AudioBase::AddDSP`. Many parameters depend on the sampling rate, but it's practically constant during gameplay, so it can be read without worrying about thread safety. However this requires passing the audio sampling rate before adding DSPs (there are 13 of them) to the track and modifying all the setters for the parameters....

------

`<rant>`

Also, while wandering around the codes running on the audio thread, I found some general problems on the structure...

First of all I don't like (at all) that all the audio is preloaded before the game is being played. However, the fix is likely to be quite complex so I'll pass.

But the more serious problem is that the codes running on the audio thread is far from real-time code.
* Dynamic allocations happen for buffering per-channel audio data.
* Lock is used for ensuring thread safety (and it failed its job).
* Even if the lock *were* necessary, the critical section seems to be too large and not efficient (especially `AudioBase::AddDSP` and `AudioBase::RemoveDSP`).

Sadly I know that I don't know enough about audio DSP, and I don't have a good idea on how to reduce the critical section, so I don't think that I'm able to fix everything properly without making someone cringy.

`</rant>`

------

TL;DR: a crash bug. annoying to fix. much more annoying to fix it properly.
